packages/@aws-cdk/aws-iam/lib/policy-document.ts

```diff
@@ -20,13 +20,40 @@ export class PolicyDocument extends cdk.Token {
       return undefined;
     }
 
-    const doc = this.baseDocument || { };
-    doc.Statement = doc.Statement || [ ];
-    doc.Version = doc.Version || '2012-10-17';
-    doc.Statement = doc.Statement.concat(this.statements);
+    const doc = {
+      ...this.baseDocument,
+      Statement: (this.baseDocument.Statement || []).concat(this.statements),
+      Version: this.baseDocument.Version || '2012-10-17'
+    };
+
     return doc;
   }
 
+  /**
+   * Removes duplicate statements
+   */
+  public postProcess(input: any, _context: cdk.ResolveContext): any {
+    if (!input || !input.Statement) {
+      return input;
+    }
+
+    const jsonStatements: string[] = [];
+    const uniqueStatements: PolicyStatement[] = [];
+
+    for (const statement of input.Statement) {
+      const jsonStatement = JSON.stringify(statement);
+      if (!jsonStatements.includes(jsonStatement)) {
```

Comment:
Make `jsonStatements` a `new Set<string>()`, so you get faster existence check with `has`.

Reply:
I think it's more a matter of preferences here, what you gain in `has` you lose in `add`, no?. Will adapt.

Reply:
I think the `O(N^2)` check is fine, `N` will be small. But I'm not opposed to using sets either. The difference will probably be a wash.